### DISCUSSION TITLE
Typechecking within parcels to provide better error messages

### DISCUSSION THREAD

**Discussion by VeckoTheGecko** [Parcels-code/Parcels#1234] [Ideas]
I think it would be good to enforce types within the user level API (eg. `if not isinstance(obj, class): raise ...`). Some error messages are very uninformative otherwise.

# Example
I wanted to combine a list of kernels down to a single combined kernel. Typing this:
```py
import functools
combined_kern = functools.reduce(lambda x, y: pset.Kernel(x) + pset.Kernel(y), [test_kern, test_kern, test_kern])
```

raised this very cryptic traceback:
```
Exception has occurred: AttributeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
'KernelSOA' object has no attribute '__name__'
  File "C:\Users\nickh\anaconda3\envs\parcels_hons\Lib\site-packages\parcels\kernel\basekernel.py", line 75, in __init__
    self.funcname = funcname or pyfunc.__name__
  File "C:\Users\nickh\anaconda3\envs\parcels_hons\Lib\site-packages\parcels\kernel\kernelsoa.py", line 47, in __init__
    super(KernelSOA, self).__init__(fieldset=fieldset, ptype=ptype, pyfunc=pyfunc, funcname=funcname, funccode=funccode, py_ast=py_ast, funcvars=funcvars, c_include=c_include, delete_cfiles=delete_cfiles)
  File "C:\Users\nickh\anaconda3\envs\parcels_hons\Lib\site-packages\parcels\particleset\particlesetsoa.py", line 650, in Kernel
    return Kernel(self.fieldset, self.collection.ptype, pyfunc=pyfunc, c_include=c_include,
  File "C:\Users\nickh\oceans\repos\honours-modelling\my_modelling\parcels_custom.py", line 213, in <lambda>
    combined_kern = functools.reduce(lambda x, y: pset.Kernel(x) + pset.Kernel(y), [test_kern, test_kern, test_kern])
  File "C:\Users\nickh\oceans\repos\honours-modelling\my_modelling\parcels_custom.py", line 213, in <module>
    combined_kern = functools.reduce(lambda x, y: pset.Kernel(x) + pset.Kernel(y), [test_kern, test_kern, test_kern])
```

The underlying error is that `pset.Kernel` is applied twice by the `lambda` (once on the left, and once on the right) and the `Kernel` method can't handle this as it does not verify that `pyfunc` is `Callable`:
https://github.com/OceanParcels/parcels/blob/7f6610417c948e85d65d98704db36e55e505b5a7/parcels/particleset/particlesetsoa.py#L644-L651

Clarifying this by having raising an error ("`pyfunc` must be Callable") or having `pset.Kernel(kern)` return `kern` would help here. In general, having this across the codebase would prevent similar errors.

Of course, my example can be solved by doing:
```py
import functools
combined_kern = functools.reduce(lambda x, y: x+y, (pset.Kernel(i) for i in [test_kern, test_kern, test_kern]))
```

**Comment by erikvansebille** [comment]
Good catch, @VeckoTheGecko. Could you add this check in the code yourself and create a PR?
